Enable the pipeline to run with inputs from multiple BIDS datasets

mpompolas opened the pull request:
## Description
<!-- describe what the PR is about. Explain the approach and possible drawbacks.It's ok to repeat some text from the related issue. -->

PR for allowing the usage of multiple BIDS datasets as input in the config.json
The field: `bids_path` will now be a list

This allows to run the pipeline without copying the files to a new dataset.

It creates a merged `participants.tsv` that gets saved in the `log_directory`.
This new .tsv also has an extra column with the folder that the subjects correspond to.
This .tsv will also be used in https://github.com/ivadomed/ivadomed/pull/572#issue-536366775 for creating model comparisons based on any metadata available.


Question: Should we also dump in the .joblib the BIDS path that the files came from for completeness?


This is a working version for multiple variations of the config parameters:
1. `command: train/test`
2.  `Film On/Off`
3. `split_dataset: Method: per_center/per_patient`
4. `split_dataset: center_test: []/["a","b","c"]`


P.S. this should not be confused with: https://github.com/ivadomed/ivadomed/pull/573 which is not part of the pipeline.


## TODO
Tests I can think of for bug discovery are:
1. Run it with Hemis enabled (probably wait until Hemis is done)
2. Run it with bounding box parameters - currently working on this
3. roi_params


Any other suggections for tests are welcome.

mpompolas committed:
WIP add multiple datasets on `config.json`
bids_path entries will be be a list instead of a string

mpompolas committed:
WIP
Working version of merging multiple BIDS datasets from selecting them from a list on config.json.

TODO: test on HEMIS
TODO: test Bounding box entries on: loader/loader.py

mpompolas committed:
Took care of the possibility that the same subject exists in more than one dataset.
Only the files within the first BIDS dataset will be used

andreanne-lemay reviewed (commented):
First pass, looks good in general. Only (more important) comment would be to add code to continue to support the current behavior, so we don't have to change every past configuration files (see review)

mpompolas commented on ivadomed/config/config.json:
```diff
@@ -9,7 +9,7 @@
     },
     "loader_parameters": {
-        "bids_path": "data_example_spinegeneric",
+        "bids_path": ["data_example_spinegeneric"],
```
Agreed

mpompolas committed:
WIP - Still need to do the testing with bounding boxes etc.

Corrections based on Andreanne's suggestions.
A string for bids_path can be assigned at config.json

charleygros commented:
Looks great to me!
and Agree with Andréanne's comment
Minor thing: could you also please update the documentation, [here](https://github.com/ivadomed/ivadomed/blob/master/docs/source/configuration_file.rst), param `bids_path`, cheers!

mpompolas committed:
Merge from Master

mpompolas committed:
Merge remote-tracking branch 'origin/master' into kn/addMultipleDatasetsConfig

# Conflicts:
#	ivadomed/config/config.json
#	ivadomed/config/config_classification.json
#	ivadomed/config/config_default.json
#	ivadomed/config/config_new_loader.json
#	ivadomed/config/config_sctTesting.json
#	ivadomed/config/config_small.json
#	ivadomed/config/config_spineGeHemis.json
#	ivadomed/config/config_tumorSeg.json
#	ivadomed/config/config_vertebral_labeling.json
#	ivadomed/evaluation.py
#	ivadomed/loader/loader.py
#	ivadomed/loader/utils.py
#	ivadomed/main.py

mpompolas committed:
Working version after merging the latest master version

mpompolas committed:
Merge branch 'master' into kn/addMultipleDatasetsConfig

mpompolas committed:
Updated docs

mpompolas committed:
Merge remote-tracking branch 'origin/master' into kn/addMultipleDatasetsConfig

mpompolas committed:
Merge remote-tracking branch 'origin/kn/addMultipleDatasetsConfig' into kn/addMultipleDatasetsConfig

mpompolas commented:
Hi @ahill187, I want to merge this PR but I'm having trouble with the tests.
Can you please take care of the testing part so we are more efficient?

ahill187 commented:
> Hi @ahill187, I want to merge this PR but I'm having trouble with the tests.
> Can you please take care of the testing part so we are more efficient?

Sure no problem!

ahill187 committed:
Fix unit tests

mariehbourget commented on ivadomed/loader/utils.py:
```diff
@@ -296,6 +297,9 @@ def get_new_subject_split_new(df, split_method, data_testing, random_seed,
     split_path = os.path.join(path_output, "split_datasets.joblib")
     joblib.dump(split_dct, split_path)
 
+    # Save the newly created .tsv on the output folder to be used during evaluation
```
Here, you save the df in the `get_new_subject_split_new` function. This function is part of the new loader project and not integrated in the pipeline yet. You may want to move the save in the `get_new_subject_split` function above where you merge the 2 datasets.

jcohenadad commented on docs/source/configuration_file.rst:
```diff
@@ -143,8 +143,8 @@ Loader Parameters
-        "description": "Path of the BIDS folder.",
-        "type": "string"
+        "description": "Path(s) of the BIDS folder(s).",
+        "type": "list"
```
could also be str (based on meeting discussion)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/b50aaeb46abfec8595237388c1c6eb4e84d9979a

jcohenadad commented on ivadomed/evaluation.py:
```diff
@@ -35,6 +35,10 @@ def evaluate(path_data, path_output, target_suffix, eval_params):
 
+    # Load participants.tsv from the output folder - This will be used to get the BIDS folder that each subject was
+    # derived from - Not using the path_data input anymore - remove dependency
+    df_participantsTSV = pd.read_table(os.path.join(path_output, 'participants.tsv'), sep=",", encoding="ISO-8859-1")
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/404e4db313f1704cbe0376a24ea00720df110c27

jcohenadad commented on ivadomed/evaluation.py:
```diff
@@ -46,8 +50,14 @@ def evaluate(path_data, path_output, target_suffix, eval_params):
-        fname_gt = [os.path.join(path_data, 'derivatives', 'labels', subj, 'anat', subj_acq + suffix + '.nii.gz')
+
+        # Use the path that is stored on the .tsv file for retrieving the derivative
+        the_BIDS_path = df_participantsTSV[df_participantsTSV['participant_id'] == subj]['path_output'].tolist()
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/404e4db313f1704cbe0376a24ea00720df110c27

jcohenadad commented on ivadomed/loader/adaptative.py:
```diff
@@ -176,7 +176,7 @@ class BIDStoHDF5:
 
     Args:
-        root_dir (str): Path to the BIDS dataset.
+        root_BIDS_list (list): List with paths to the BIDS dataset(s).
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/7530f45f452000a990a841167d569bab5f527f92

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/0a858ef50e35632d3e0473179121bbaded3ff81c

jcohenadad commented on ivadomed/loader/adaptative.py:
```diff
@@ -203,13 +203,21 @@ class BIDStoHDF5:
+            self.bids_ds.append(bids.BIDS(BIDSFolder))
+        # Append subjects from all BIDSdatasets into a list
+        bids_subjects = [s for s in self.bids_ds[0].get_subjects() if s.record["subject_id"] in subject_lst]
+        for iBIDSFolder in range(1, len(root_BIDS_list)):
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/214f78d19437357e16d22d20aa7ef78f0536c09a

jcohenadad commented on ivadomed/loader/adaptative.py:
```diff
@@ -236,9 +244,14 @@ def __init__(self, root_dir, subject_lst, target_suffix, contrast_lst, path_hdf5
 
+        # Append get_subjects()
+        get_subjects_all = self.bids_ds[0].get_subjects()
+        for iBIDSFolder in range(1, len(self.bids_ds)):
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/214f78d19437357e16d22d20aa7ef78f0536c09a

jcohenadad commented on ivadomed/loader/loader.py:
```diff
@@ -772,19 +774,25 @@ class BidsDataset(MRI2DSegmentationDataset):
-        bids_subjects = [s for s in self.bids_ds.get_subjects() if s.record["subject_id"] in subject_lst]
+        # Append subjects from all BIDSdatasets into a list
+        bids_subjects = [s for s in self.bids_ds[0].get_subjects() if s.record["subject_id"] in subject_lst]
+        for iBIDSFolder in range(1, len(root_dirs)):
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/214f78d19437357e16d22d20aa7ef78f0536c09a

jcohenadad commented on ivadomed/loader/loader.py:
```diff
@@ -810,10 +818,15 @@ def __init__(self, root_dir, subject_lst, target_suffix, contrast_params, slice_
 
+        # Append get_subjects()
+        get_subjects_all = self.bids_ds[0].get_subjects()
+        for iBIDSFolder in range(1, len(self.bids_ds)):
```
PEP8 uses snake_case (no capital letters)

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/214f78d19437357e16d22d20aa7ef78f0536c09a

jcohenadad commented on ivadomed/loader/utils.py:
```diff
@@ -722,6 +726,69 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
     return nib.orientations.apply_orientation(arr_ras, trans_orient)
 
 
+def merge_bids_datasets(BIDS_dataset_paths_list):
```
if `BIDS_dataset_paths_list` is a variable should be all small caps (snake_case)-- pls correct everywhere

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/b35ade046cc9e65e574f45f197867d5d73f77240

jcohenadad commented on ivadomed/loader/adaptative.py:
```diff
@@ -176,7 +176,7 @@ class BIDStoHDF5:
 
     Args:
-        root_dir (str): Path to the BIDS dataset.
+        root_BIDS_list (list): List with paths to the BIDS dataset(s).
```
why not accepting here too str and list (to avoid all this refactoring). This could be done with a decorator i guess. @ahill187 @gisellemartel can chip in

ahill187 replied:
Yah good point!

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/7530f45f452000a990a841167d569bab5f527f92

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/0a858ef50e35632d3e0473179121bbaded3ff81c

ahill187 committed:
Merge branch 'master' into kn/addMultipleDatasetsConfig

ahill187 committed:
convert df_participantsTSV to snake_case

ahill187 committed:
Merge branch 'kn/addMultipleDatasetsConfig' of https://github.com/ivadomed/ivadomed into kn/addMultipleDatasetsConfig

ahill187 committed:
covert iBIDSFolder to snake_case

ahill187 committed:
Refactor path_data backwards compatibility

Create function `format_path_data` in `utils.py`

ahill187 committed:
Refactor loader.py to allow path_data as list/str

Refactor BidsDataset and Bids3DDataset to rename parameter root_dirs
to path_data, allow path_data as list or str.

Update test_orientation from root_dirs to path_data.

ahill187 committed:
rename path_folders to path_data for consistency

ahill187 committed:
rename BIDS_dataset_paths_list to path_data, allow for str

ahill187 committed:
convert iDataset to i_dataset snake_case

ahill187 committed:
Change datasets logging to allow for list

ahill187 committed:
update config file docs for path_data list and string

ahill187 committed:
Remove f.close() typo/bug in main

ahill187 committed:
update adaptative.py to allow list or str for path_data

ahill187 committed:
fix path_data renaming

ahill187 committed:
allow path_data CLI to accept list

jcohenadad commented on ivadomed/loader/adaptative.py:
```diff
@@ -467,7 +482,7 @@ class HDF5Dataset:
 
     Args:
-        root_dir (path): Path of bids and data.
+        path_data (list) or (str): Path of bids and data.
```
```suggestion
        path_data (list) or (str): Path(s) to BIDS datasets
```

ahill187 replied:
Fixed here: https://github.com/ivadomed/ivadomed/pull/588/commits/1dfbe939fadc23bee2cf1ea345d7244f1117f848

ahill187 committed:
Merge branch 'master' into kn/addMultipleDatasetsConfig

ahill187 committed:
Change description for path_data in adaptative.py

jcohenadad commented:
thank you for the fixes @ahill187 ! and thank you for introducing this awesome feature @mpompolas 🙏

gisellemartel commented on ivadomed/loader/adaptative.py:
```diff
@@ -176,7 +177,7 @@ class BIDStoHDF5:
 
     Args:
-        root_dir (str): Path to the BIDS dataset.
+        path_data (list) or (str): Path(s) to BIDS datasets
```
using `list` as a variable name could be a naming conflict with the  built-in python `list()`. I would suggest maybe using `lst` instead?
